Automatic detection of record format

lutovich opened the pull request:
This commit removes the need to always explicitly configure record format, if it is something else than default. This means that previously, tools like consistency checked and backup always needed an explicit config parameter `GraphDatabaseSettings.record_format` when applied to a high_limit store.

Detection of format happens inside `NeoStoreDataSource` and `StoreFactory`. Later is often directly used by tools like consistency checker.

lutovich committed:
Automatic detection of record format

This commit removes the need to always explicitly configure record format, if
it is something else than default. This means that previously, tools like
consistency checked and backup always needed an explicit config parameter
`GraphDatabaseSettings.record_format` when applied to a high_limit store.

Detection of format happens inside `NeoStoreDataSource` and `StoreFactory`.
Later is often directly used by tools like consistency checker.

MishaDemianenko commented:
@lutovich it looks like this PR should be included into change logs?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/internal/BatchInserterImpl.java:
```diff
@@ -248,11 +245,10 @@ public BatchInserterImpl( final File storeDir, final FileSystemAbstraction fileS
-                recordFormats, logService.getInternalLogProvider() );
+                logService.getInternalLogProvider() );
+
+        maxNodeId = sf.getRecordFormats().node().getMaxId();
```
In this example storeFactory is used as record format provider, but it should not be; it's not a storeFactory responsibility

MishaDemianenko commented on enterprise/neo4j-enterprise/src/test/java/org/neo4j/consistency/ConsistencyCheckServiceRecordFormatIT.java:
```diff
@@ -0,0 +1,141 @@
+
+    @Rule
+    public final RuleChain ruleChain = RuleChain.outerRule( SuppressOutput.suppressAll() )
+            .around( SystemExitRule.none() )
```
why do we need SystemExitRule here?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/NeoStoresTest.java:
```diff
@@ -521,9 +519,8 @@ public void shouldNotReadNonRecordDataAsRecord() throws Exception
-                NullLogProvider.getInstance() );
-        long recordVersion = MetaDataStore.versionStringToLong( getRecordFormat().storeVersion() );
+        StoreFactory factory = new StoreFactory( neoStoreDir, pageCache, fileSystem, NullLogProvider.getInstance() );
+        long recordVersion = MetaDataStore.versionStringToLong( factory.getRecordFormats().storeVersion() );
```
here also store factory used as record formats source

MishaDemianenko commented on enterprise/kernel/src/test/java/org/neo4j/kernel/impl/store/format/RecordFormatSelectorTest.java:
```diff
@@ -21,70 +21,282 @@
+        }
+        catch ( Exception e )
+        {
+            assertThat( e, instanceOf( IllegalArgumentException.class ) );
```
ExpectedException rule instead?

MishaDemianenko commented on enterprise/kernel/src/test/java/org/neo4j/kernel/impl/store/format/RecordFormatSelectorTest.java:
```diff
@@ -21,70 +21,282 @@
+        }
+        catch ( Exception e )
+        {
+            assertThat( e, instanceOf( IllegalArgumentException.class ) );
```
ExpectedException rule instead?

MishaDemianenko commented on enterprise/kernel/src/test/java/org/neo4j/kernel/impl/store/format/RecordFormatSelectorTest.java:
```diff
@@ -21,70 +21,282 @@
+        }
+        catch ( Exception e )
+        {
+            assertThat( e, instanceOf( IllegalArgumentException.class ) );
```
ExpectedException rule?

MishaDemianenko commented on enterprise/kernel/src/test/java/org/neo4j/kernel/impl/store/format/RecordFormatSelectorTest.java:
```diff
@@ -21,70 +21,282 @@
+        }
+        catch ( Exception e )
+        {
+            assertThat( e, instanceOf( IllegalArgumentException.class ) );
```
ExpectedException rule?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreFactory.java:
```diff
@@ -183,4 +158,9 @@ public NeoStores openNeoStores( boolean createStoreIfNotExists, StoreType... sto
                 fileSystemAbstraction, recordFormats, createStoreIfNotExists, storeTypes );
     }
+
+    public RecordFormats getRecordFormats()
```
i don't think store factory should provide it to outside world

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/StoreVersions.java:
```diff
@@ -36,27 +38,43 @@
     // Enterprise formats
     public static final String HIGH_LIMIT_V3_0 = "vE.H.0";
 
+    private static final List<String> communityStoreVersions = Arrays.asList(
```
Now those 2 list will need to be always updated when new format will be introduced and potentially can cause fragile behaviour.
Also it has nothing to do with this PR

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -534,6 +533,34 @@ public void start() throws IOException
+    private static RecordFormats findDesiredStoreFormats( Config config, File storeDir, FileSystemAbstraction fs,
+            PageCache pageCache, LogService logService )
+    {
+        RecordFormats result;
```
Record format selection logic should not be here, it should be moved to RecordFormatSelector

MishaDemianenko commented:
@lutovich this PR removes possibility of non community edition to use any other format then default without using configuration property. This does not look like a right thing todo, especially if that was/will be desired behaviour.

MishaDemianenko commented:
@lutovich review round completed, couple of comments

lutovich commented:
@MishaDemianenko fixed all comments except `ExpectedException` because I do not like it :)
With this PR all editions have `standard` format as default. This is acceptable for 3.0 but might change in 3.1. I guess, this PR can be forward merged and if we decide to make `high_limit` enterprise default in 3.1 than all editions will need to define default format themselves. I can issue such PR after the merge, if needed.

MishaDemianenko commented:
@lutovich would it make more sense to null merge it into 3.1 and create another pr for that version?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/LabelTokenStoreTest.java:
```diff
@@ -74,7 +73,7 @@ public void getRecord() throws IOException
         {
             super( file, config, generatorFactory, cache, logProvider, dynamicStringStore,
-                    autoSelectFormat( config, NullLogService.getInstance() ) );
+                    StandardV3_0.RECORD_FORMATS );
```
should it still be any format that is default and not StandardV3_0 always?

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/test/NeoStoresRule.java:
```diff
@@ -62,18 +60,14 @@ public NeoStoresRule( Class<?> testClass, StoreType... stores )
-        Config conf = new Config( stringMap( config ) );
-        return open( RecordFormatSelector.select( conf, RecordFormatSelector.autoSelectFormat(),
-                NullLogService.getInstance() ), config );
+        return open( StandardV3_0.RECORD_FORMATS, config );
```
why not autoselected?

lutovich committed:
Auto-select record format for tests

Commit makes couple tests use default format and auto-select format based on
config instead of hardcoded StandardV3_0.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseRecordFormats.java:
```diff
@@ -35,16 +35,24 @@
 
     @Override
+    public String name()
+    {
```
also as we have already talk, maybe we do not need name to be exposed here

MishaDemianenko commented:
@lutovich couple of small comments, and we need to decide how we will merge it into 3.1 before proceed

lutovich committed:
Removed RecordFormats#name()

This method was used for logging and to compare configured format with the
store format. Format class names work just fine for logging and format
comparison changed to use RecordFormats#generation();

lutovich commented:
Build fails because of the issue fixed here: https://github.com/neo4j/neo4j/pull/7260.
This PR uses record format generations and they are a bit off. It is probably reasonable to wait for https://github.com/neo4j/neo4j/pull/7260 and merge this one afterwards.

MishaDemianenko commented:
@lutovich can you please rebase it, #7260 has been merged.

MishaDemianenko commented:
@lutovich some test failures

lutovich commented:
@MishaDemianenko rebase did not go well... should be fixed now
